lib/types.js

```diff
@@ -533,77 +533,119 @@ module.exports = function () {
+                  " does not match field " + field +
+                  " of type " + self.typeName
+                );
+            }
 
-                self.buildParams.forEach(function (param, i) {
-                    add(param, i);
-                });
+            built[param] = value;
+        }
 
-                Object.keys(self.allFields).forEach(function (param) {
-                    add(param); // Use the default value.
-                });
+        // Calling the builder function will construct an instance of the Def,
+        // with positional arguments mapped to the fields original passed to .build.
+        // If not enough arguments are provided, the default value for the remaining fields
+        // will be used.
+        function builder() {
+            var args = arguments;
+            var argc = args.length;
+            var built = Object.create(nodePrototype);
+
+            if (!self.finalized) {
+                throw new Error(
+                  "attempting to instantiate unfinalized type " +
+                  self.typeName
+                );
+            }
 
-                // Make sure that the "type" field was filled automatically.
-                if (built.type !== self.typeName) {
-                    throw new Error("");
+            self.buildParams.forEach(function (param, i) {
+                if (i < argc) {
+                    addParam(built, param, args[i], true)
+                } else {
+                    addParam(built, param, null, false);
                 }
+            });
 
-                return built;
+            Object.keys(self.allFields).forEach(function (param) {
+                // Use the default value.
+                addParam(built, param, null, false);
+            });
+
+            // Make sure that the "type" field was filled automatically.
+            if (built.type !== self.typeName) {
+                throw new Error("");
             }
+
+            return built;
+        }
+
+        // Calling .from on the builder function will construct an instance of the Def,
+        // using field values from the passed object. For fields missing from the passed object,
+        // their default value will be used.
+        builder.from = function (obj) {
+            var built = Object.create(nodePrototype);
+
+            if (!self.finalized) {
+                throw new Error(
+                  "attempting to instantiate unfinalized type " +
+                  self.typeName
+                );
+            }
+
+            self.buildParams.forEach(function (param) {
+                if (hasOwn.call(obj, param)) {
+                    addParam(built, param, obj[param], true);
+                } else {
+                    addParam(built, param, null, false);
+                } 
+            });
+
+            Object.keys(self.allFields).forEach(function (param) {
+                // Use the default value.
+                addParam(built, param, null, false);
```

Comment:
I think this loop also needs to check whether `obj[param]` is defined, so that `obj` can provide values for more than just the `.build` parameters, e.g.
```js
builders.identifier.from({ name: "foo", typeAnnotation: ... })
```
The `typeAnnotation` parameter isn't one of the `.build` parameters, but it seems like it should get checked and included, if provided.

Reply:
Ah, you're right. I think we don't even need to loop over `self.buildParams` either since any field in `self.allFields` could be specified in `obj`. Now updated.